src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -1973,27 +1972,32 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
             }
             else
             {
-                rewriteAndCache();
+                rewriteAndCache(diagnosticBag);
             }
 
-            void bind(CSharpSyntaxNode bindableRoot, out Binder binder, out BoundNode boundRoot)
+            void bind(CSharpSyntaxNode bindableRoot, DiagnosticBag diagnosticBag, out Binder binder, out BoundNode boundRoot)
             {
                 binder = GetEnclosingBinder(GetAdjustedNodePosition(bindableRoot));
-                boundRoot = Bind(binder, bindableRoot, getDiagnosticBag());
+                boundRoot = Bind(binder, bindableRoot, diagnosticBag);
             }
 
-            void rewriteAndCache()
+            void rewriteAndCache(DiagnosticBag diagnosticBag)
             {
-                boundRoot = RewriteNullableBoundNodesWithSnapshots(boundRoot, binder, getDiagnosticBag(), takeSnapshots: true, out snapshotManager, ref remappedSymbols);
+#if DEBUG
+                if (!Compilation.NullableSemanticAnalysisEnabled)
+                {
+                    AnalyzeBoundNodeNullability(boundRoot, binder, diagnosticBag, takeSnapshots: true);
```

Comment:
I understand that `diagnosticBag` here will be a freshly created bag whenever we hit this code. This still feels very indirect though. Have to think a bit to verify this code isn't potentially adding new diagnostics in `DEBUG` mode. Is there a reason to not just use `new DiagnosticBag()` here to make this clearer?

Reply:
I wanted to avoid having the debug verification step create multiple throwaway diagnostic bags.

---
In reply to: [443950238](https://github.com/dotnet/roslyn/pull/42403#discussion_r443950238) [](ancestors = 443950238)